File: Tree/C++/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> res;
    
    vector<vector<int>> levelOrder(TreeNode* root) {
        helper(root, 0);
        return res;
    }
    
    void helper(TreeNode* root, int level) {
        if (root == NULL) {
            return;
        }
        if (level == res.size()) {
            res.push_back(vector<int>());
        }
        res[level].push_back(root->val);
        helper(root->left, level+1);
        helper(root->right, level+1);
    }
};

File: Microsoft/Golang/binary-tree-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    res := make([][]int, 0)
    if root == nil {
        return res
    }
    queue := make([]*TreeNode, 0)
    queue = append(queue, root)
    for len(queue) > 0 {
        n := len(queue)
        level := make([]int, 0)
        for i := 0; i < n; i++ {
            node := queue[i]
            level = append(level, node.Val)
            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }
        res = append(res, level)
        queue = queue[n:]
    }
    return res
}
